spec/controllers/campaign_pages_controller_spec.rb

```diff
@@ -1,28 +1,77 @@
 require 'rails_helper'
-require 'capybara/poltergeist'
-require 'helper_functions'
-
-RSpec.describe CampaignPagesController, type: :feature do
-  scenario 'Redirect after a user signs the petition' do
-    cam_page = create_petition_page
-    visit "/campaign_pages/#{cam_page.id}"
-    expect(find('#redirect-location')).to have_text 'www.google.com'
-  end
 
-  scenario 'Index displays the list of active campaign pages' do
-    cam_page = create_petition_page
-    log_in
-    visit '/campaign_pages'
-    expect(page.body).to have_text cam_page.title
-  end
+RSpec.describe CampaignPagesController, type: :controller do
+
+  let(:english) { create :language }
+  let( :admin ) { create :admin }
+  let(:petition_widget_params) { attributes_for :petition_widget }
+  let(:text_widget_params_1) { attributes_for :text_widget }
+  let(:text_widget_params_2) { attributes_for :text_widget }
+  let(:widget_params) { [petition_widget_params, text_widget_params_1, text_widget_params_2] }
+  let(:page_params) { attributes_for :widgetless_page, language_id: english.id }
+  let(:page_widget_params) { page_params.merge({widgets_attributes: widget_params}) }
+
+  # for error cases
+  let(:bad_widget_params) { attributes_for :text_widget, content: { body_html: nil } }
+  let(:bad_page_params) { attributes_for :widgetless_page, language_id: nil }
+  let(:bad_widget_page_params) { page_params.merge({widgets_attributes: bad_widget_params}) }
+  let(:widget_bad_page_params) { bad_page_params.merge({widgets_attributes: widget_params}) }
+
+  # let(:existing_page) { p = CampaignPage.new(page_widget_params); p.save!; p }
+
+  describe "logged in as admin" do
+
+    before :each do
+      expect(admin).to be_persisted
+      allow(controller).to receive(:current_user) { admin }
+    end
+
+    describe 'create' do
+
+      describe 'success' do
+
+        after :each do
+          expect(response).to redirect_to CampaignPage.last
+        end
+
+        it 'should be able to create a page without widgets' do
+          expect{ post :create, campaign_page: page_params }.to change{ CampaignPage.count }.by 1
```

Comment:
I'm curious what this test code would look like refactored as controller unit tests. Each test here calls a method through http, and then measures its side effects. How do we change that? Do we still get to do things like check where it redirects or renders? That seems like a valuable parts of controller testing.